Add replaceInstrument() to Plugin API

manolo opened the pull request:
Resolves: #31201 <!-- Replace `NNNNN` with a GitHub issue number, or a direct link if the issue is not on GitHub -->

<!-- Add a short description of and motivation for the changes here -->

Add a new `Score.replaceInstrument(part, instrumentId)` method to the Plugin API, allowing plugins to programmatically change instruments on score parts.

  ## Changes
  - Add `Q_INVOKABLE Score::replaceInstrument(Part*, QString)` to the plugin API
  - Bridge plugin API to existing `NotationParts::replaceInstrument()` infrastructure
  - Use `Part::MAIN_INSTRUMENT_TICK` to correctly identify the main instrument vs mid-score instrument changes
  - Add unit tests for instrument replacement with undo/redo support
  - Add test environment setup for API tests

  ## Usage Example
  
  ```javascript
  var part = curScore.parts[0];
  curScore.replaceInstrument(part, "violin");
  ```

  The instrumentId parameter should match IDs from `instruments.xml`.

----

<!-- Replace `[ ]` with `[x]` to fill the checkboxes below -->

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/muse_framework/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [x] I created a unit test or vtest to verify the changes I made (if applicable)

manolo committed:
Add replaceInstrument() to Plugin API

Expose instrument replacement functionality to plugins through new
Score.replaceInstrument() method. Allows plugins to change a part's
instrument (name, clef, transposition, sound) with full undo/redo support.

Changes:
- Add Q_INVOKABLE Score::replaceInstrument(Part*, QString) to plugin API
- Use Part::MAIN_INSTRUMENT_TICK (not Fraction(0,1)) to correctly identify
  main instrument vs. mid-score instrument changes

Usage example:
  var part = curScore.parts[0];
  curScore.replaceInstrument(part, "violin");

manolo committed:
Simplify replaceInstrument() to use DOM-level ChangePart directly

Remove dependency on notation module by using engraving::ChangePart
command directly instead of going through INotationParts. This:
- Eliminates need for NotationMock and NotationPartsMock in tests
- Simplifies test setup (no mock injection required)
- Keeps the implementation purely at the engraving layer
- Makes the API self-contained within the plugin API module

manolo committed:
Refactor replaceInstrument to share logic between NotationParts and Plugin API

Move instrument replacement logic to editpart.cpp so that both
NotationParts and the Plugin API use the same implementation.

- Add replacePartInstrument() for main instrument replacement
- Add replaceInstrumentAtTick() for instrument changes mid-score
- Update NotationParts::replaceInstrument to use new shared functions
- Update Plugin API Score::replaceInstrument to use new shared functions

cbjeukendrup commented on src/engraving/editing/editpart.h:
```diff
@@ -203,4 +203,20 @@ class SetUserBankController : public UndoCommand
+
+/// Replaces the main instrument for a part.
+/// Updates the part name, instrument, and clefs for all staves.
+void replacePartInstrument(Score* score, Part* part, const Instrument& newInstrument, const StaffType* newStaffType = nullptr,
```
We might want to put these functions as static methods in an `EditPart` class; somehow that feels more structured to me than a bunch of free functions. But it's not a big deal; if you think there are reasons to not do that, then we should not do it.

manolo committed:
Address review suggestions

- Use EditPart class with static methods instead of free functions
- Add ChangePart::cleanup() to fix instrument pointer leak
- Remove unused NotationParts::findInstrumentChange
- Replace QString with u"..." literals in tests

manolo commented on src/notation/internal/notationparts.cpp:
```diff
@@ -717,33 +717,13 @@ void NotationParts::replaceInstrument(const InstrumentKey& instrumentKey, const
+        String newInstrumentPartName = formatInstrumentTitle(newInstrument.trackName(), newInstrument.trait());
+        mu::engraving::replacePartInstrument(score(), part, newInstrument, newStaffType, newInstrumentPartName);
     } else {
-        mu::engraving::InstrumentChange* instrumentChange = findInstrumentChange(part, instrumentKey.tick);
```
done

manolo replied:
done

manolo commented on src/engraving/api/tests/score_tests.cpp:
```diff
@@ -0,0 +1,255 @@
+    initialInstrument.setTrackName(u"Flute");
+    domPart->setInstrument(initialInstrument);
+
+    EXPECT_EQ(domPart->instrumentId(), QString("test.flute"));
```
done

zacjansheski commented:
Tested on MacOS Tahoe 26.2, Windows 11, Ubuntu 22.04.3. Approved
https://github.com/musescore/MuseScore/issues/31201 FIXED
